net/portmapper/portmapper.go

```diff
@@ -0,0 +1,601 @@
+
+	prevPort := uint16(0) // or 0 if unknown
+
+	// Do we have an existing mapping that's valid?
+	now := time.Now()
+	if m := c.pmpMapping; m != nil {
+		if now.Before(m.useUntil) {
+			defer c.mu.Unlock()
+			return m.external, nil
+		}
+		// The mapping might still be valid, so just try to renew it.
+		prevPort = m.external.Port
+	}
+
+	// If we just did a Probe (e.g. via netchecker) but didn't
+	// find a PMP service, bail out early rather than probing
+	// again. Cuts down latency for most clients.
+	haveRecentPMP := c.sawPMPRecentlyLocked()
+	if haveRecentPMP {
+		m.external.IP = c.pmpPubIP
+	}
+	if c.lastProbe.After(now.Add(-5*time.Second)) && !haveRecentPMP {
+		return netaddr.IPPort{}, NoMappingError{ErrNoPortMappingServices}
+	}
+
+	c.mu.Unlock()
+
+	uc, err := netns.Listener().ListenPacket(ctx, "udp4", ":0")
+	if err != nil {
+		return netaddr.IPPort{}, err
+	}
+	defer uc.Close()
+
+	uc.SetReadDeadline(time.Now().Add(portMapServiceTimeout))
+	defer closeCloserOnContextDone(ctx, uc)()
+
+	pmpAddr := netaddr.IPPort{IP: gw, Port: pmpPort}
+	pmpAddru := pmpAddr.UDPAddr()
+
+	// Ask for our external address if needed.
+	if m.external.IP.IsZero() {
+		if _, err := uc.WriteTo(pmpReqExternalAddrPacket, pmpAddru); err != nil {
+			return netaddr.IPPort{}, err
+		}
+	}
+
+	// And ask for a mapping.
+	pmpReqMapping := buildPMPRequestMappingPacket(localPort, prevPort, pmpMapLifetimeSec)
+	if _, err := uc.WriteTo(pmpReqMapping, pmpAddru); err != nil {
+		return netaddr.IPPort{}, err
+	}
+
+	res := make([]byte, 1500)
+	for {
+		if m.externalValid() {
+			c.mu.Lock()
+			defer c.mu.Unlock()
+			c.pmpMapping = m
+			return m.external, nil
+		}
+		n, srci, err := uc.ReadFrom(res)
+		if err != nil {
+			if ctx.Err() == context.Canceled {
+				return netaddr.IPPort{}, err
+			}
+			return netaddr.IPPort{}, NoMappingError{ErrNoPortMappingServices}
+		}
+		srcu := srci.(*net.UDPAddr)
+		src, ok := netaddr.FromStdAddr(srcu.IP, srcu.Port, srcu.Zone)
+		if !ok {
+			continue
+		}
+		if src == pmpAddr {
+			pres, ok := parsePMPResponse(res[:n])
+			if !ok {
+				c.logf("unexpected PMP response: % 02x", res[:n])
+				continue
+			}
+			if pres.ResultCode != 0 {
+				return netaddr.IPPort{}, NoMappingError{fmt.Errorf("PMP response Op=0x%x,Res=0x%x", pres.OpCode, pres.ResultCode)}
+			}
+			if pres.OpCode == pmpOpReply|pmpOpMapPublicAddr {
+				m.external.IP = pres.PublicAddr
+			}
+			if pres.OpCode == pmpOpReply|pmpOpMapUDP {
+				m.external.Port = pres.ExternalPort
+				d := time.Duration(pres.MappingValidSeconds) * time.Second
+				d /= 2 // renew in half the time
+				m.useUntil = time.Now().Add(d)
+				m.epoch = pres.SecondsSinceEpoch
+			}
+		}
+	}
+}
+
+type pmpResultCode uint16
+
+// NAT-PMP constants.
+const (
+	pmpPort              = 5351
+	pmpMapLifetimeSec    = 7200 // RFC recommended 2 hour map duration
+	pmpMapLifetimeDelete = 0    // 0 second lifetime deletes
+
+	pmpOpMapPublicAddr = 0
+	pmpOpMapUDP        = 1
+	pmpOpReply         = 0x80 // bit
```

Comment:
not convinced // bit adds much value here, but if it's conventional, wfm.

Reply:
clarified comment in both places